Cesium3DTilesSelection/src/Tileset.cpp

```diff
@@ -612,68 +633,105 @@ static bool isVisibleInFog(double distance, double fogDensity) noexcept {
+      return;
+    }
+    // Frustum cull based on the actual tile's bounds.
+  } else if (std::any_of(
+                 frustums.begin(),
+                 frustums.end(),
+                 [&boundingVolume = tile.getBoundingVolume(),
+                  renderTilesUnderCamera =
+                      this->_options.renderTilesUnderCamera](
+                     const ViewState& frustum) {
+                   return isVisibleFromCamera(
+                       frustum,
+                       boundingVolume,
+                       renderTilesUnderCamera);
+                 })) {
+    // The tile is visible in at least one frustum, so don't cull.
+    return;
+  }
 
-  // whether we should visit this tile
-  bool shouldVisit = true;
-  // whether this tile was culled (Note: we might still want to visit it)
-  bool culled = false;
+  // If we haven't returned yet, this tile is frustum culled.
+  cullResult.culled = true;
 
-  for (const std::shared_ptr<ITileExcluder>& pExcluder :
-       this->_options.excluders) {
-    if (pExcluder->shouldExclude(tile)) {
-      culled = true;
-      shouldVisit = false;
-      break;
-    }
+  if (this->_options.enableFrustumCulling) {
+    // frustum culling is enabled so we shouldn't visit this off-screen tile
+    cullResult.shouldVisit = false;
+  }
+}
+
+void Tileset::_fogCull(
+    const FrameState& frameState,
+    const std::vector<double>& distances,
+    CullResult& cullResult) {
+
+  if (!cullResult.shouldVisit || cullResult.culled) {
+    return;
   }
 
   const std::vector<ViewState>& frustums = frameState.frustums;
   const std::vector<double>& fogDensities = frameState.fogDensities;
 
-  const BoundingVolume& boundingVolume = tile.getBoundingVolume();
-  if (std::none_of(
-          frustums.begin(),
-          frustums.end(),
-          [boundingVolume,
-           renderTilesUnderCamera = this->_options.renderTilesUnderCamera](
-              const ViewState& frustum) {
-            return isVisibleFromCamera(
-                frustum,
-                boundingVolume,
-                renderTilesUnderCamera);
-          })) {
-    // this tile is off-screen so it is a culled tile
-    culled = true;
-    if (this->_options.enableFrustumCulling) {
-      // frustum culling is enabled so we shouldn't visit this off-screen tile
-      shouldVisit = false;
+  bool isFogCulled = true;
+
+  for (size_t i = 0; i < frustums.size(); ++i) {
+    const double distance = distances[i];
+    const double fogDensity = fogDensities[i];
+
+    if (isVisibleInFog(distance, fogDensity)) {
+      isFogCulled = false;
+      break;
     }
   }
 
+  if (isFogCulled) {
+    // this tile is occluded by fog so it is a culled tile
+    cullResult.culled = true;
+    if (this->_options.enableFogCulling) {
+      // fog culling is enabled so we shouldn't visit this tile
+      cullResult.shouldVisit = false;
+    }
+  }
+}
+
+CesiumUtility::ScopeGuard<std::function<void()>>
```

Comment:
I think returning a ScopeGuard here is pretty unusual. And it requires use of std::function, which is less efficient (not that it makes a big difference, probably). I think if you want to refactor this code so that the lifetime can't be managed by a scope anymore, then it shouldn't use ScopeGuard. Maybe instead use a small class that allocates a vector of distances from the pool, computes distances, and returns the vector to the pool when it goes out of scope.